### PR TITLE
Enable tray icon on Linux without breaking Windows

### PR DESCRIPTION
**Pull request by ConorIA**
Here is a suggestion to resolve issue #22, and offer a more elegant solution to issue #1. The png icons are of my own making, having traced the ico files in inkscape. They can and maybe should be replaced by official icons, but they do the trick for now!

### CODE REVIEW

File: electron/js/menu/tray.js [hunk 1]
```diff
@@ -24,14 +24,22 @@ const {app, BrowserWindow, Menu, Tray} = require('electron');
 const path = require('path');
 const config = require('./../config');
 
-const iconPath = path.join(app.getAppPath(), 'img', 'tray.ico');
-const iconBadgePath = path.join(app.getAppPath(), 'img', 'tray.badge.ico');
+var iconExt = null;
```
Comment (lipis) [inline electron/js/menu/tray.js]: `var iconExt = 'png'`

File: electron/js/menu/tray.js [hunk 2]
```diff
@@ -24,14 +24,22 @@ const {app, BrowserWindow, Menu, Tray} = require('electron');
...
+if (process.platform === 'win32') {
+  iconExt = 'ico'
```
Comment (lipis) [inline electron/js/menu/tray.js]: semicolon

File: electron/js/menu/tray.js [hunk 3]
```diff
@@ -24,14 +24,22 @@ const {app, BrowserWindow, Menu, Tray} = require('electron');
...
+}
+else {
+  iconExt = 'png'
+}
```
Comment (lipis) [inline electron/js/menu/tray.js]: no need for the else part

File: electron/js/menu/tray.js [hunk 4]
```diff
@@ -24,14 +24,22 @@ const {app, BrowserWindow, Menu, Tray} = require('electron');
...
+
+const iconPath = path.join(app.getAppPath(), 'img', ('tray.'+iconExt));
```
Comment (lipis) [inline electron/js/menu/tray.js]: spaces before and after `+`

File: electron/js/menu/tray.js [hunk 5]
```diff
@@ -24,14 +24,22 @@ const {app, BrowserWindow, Menu, Tray} = require('electron');
...
+const iconBadgePath = path.join(app.getAppPath(), 'img', ('tray.'+iconExt));
```
Comment (lipis) [inline electron/js/menu/tray.js]: `tray.badge.`

**Comment by lipis** [comment]
You should also update this part: https://github.com/wireapp/wire-desktop/blob/master/electron/js/util.js#L59-L69

to something like:

```javascript
if (process.platform === 'darwin') {
  app.dock.setBadge(count ? count[1] : '');
} else if (count) {
  tray.useBadgeIcon();
} else {
  tray.useDefaultIcon();
}
```

**Comment by lipis** [comment]
Having done all that.. it doesn't work on my Ubuntu 16.04 :/ Can you send us a screenshot wiht yours?

File: electron/js/menu/tray.js (hunk 3 above)
Comment (lipis) [inline electron/js/menu/tray.js]: Even better...  :)

**Comment by ConorIA** [comment]
I'm out of town / away from my laptop today. I'll send a screenshot over as soon as I can.

**Comment by lipis** [comment]
🌴 🍺 🍸

**Comment by ConorIA** [comment]
There's the icon with the context menu open. I use the 'top icons' extension on Gnome, so on other systems it will be in a tray on the bottom left corner. This is using the build based on my PR. I will try with the changes you have suggested. 

screenshot from 2016-08-30 16-56-43

**Comment by lipis** [comment]
Your next task is to make the travis tests green.. and we'll ask our designers to provide the correct PNGs :)

**Comment by ConorIA** [comment]
Getting a type error now. A little over my head. Any suggestions? Also, if you want to ask your designers for an SVG app icon to go with the PNG tray icons, that's be great! :smiley:

**Comment by ConorIA** [comment]
Played around a little, but still no luck. I tried replacing line 33 (`let  appIcon = null;`) with `var appIcon = new Tray(iconPath);`, but I get an error: 

```
TypeError: Error processing argument at index 0, conversion failure from /home/conor/git/wire-desktop/node_modules/electron-prebuilt/dist/resources/default_app.asar/img/tray.png
```

**Comment by bennyn** [comment]
> TypeError: Cannot read property 'setImage' of null
    at Object.useBadgeIcon (/home/travis/build/wireapp/wire-desktop/electron/js/menu/tray.js:63:12)

It looks like `appIcon` isn't defined?

**Comment by ConorIA** [comment]
As far as I can tell, the app icon is originally set to null but should be replaced as long as we're not on Darwin. Not sure why it isn't getting replaced later on. I'm not familiar with JS, so I might be reading it wrong.

**Comment by ConorIA** [comment]
Could it be that the test is simply timing out before the badge updates? On my machine, when I run `npm test --verbose`, the window barely opens before it is closed and warns: 

```
  util
    #updateBadge()
TypeError: Cannot read property 'setImage' of null
    at Object.useBadgeIcon (/home/conor/wire-desktop/electron/js/menu/tray.js:63:12)
    at Timeout._onTimeout (/home/conor/wire-desktop/electron/js/util.js:62:14)
    at tryOnTimeout (timers.js:224:11)
    at Timer.listOnTimeout (timers.js:198:5)
TypeError: Cannot read property 'setImage' of null
    at Object.useBadgeIcon (/home/conor/wire-desktop/electron/js/menu/tray.js:63:12)
    at Timeout._onTimeout (/home/conor/wire-desktop/electron/js/util.js:62:14)
    at tryOnTimeout (timers.js:224:11)
    at Timer.listOnTimeout (timers.js:198:5)
      1) should update badge according to window title
      ✓ should update badge according to window title (1498ms)
```

I have tried to adjust the timeout with `this.timeout`, but it seems that the use of arrow functions breaks that option (https://mochajs.org/; https://github.com/mochajs/mocha/issues/2018).

**Comment by lipis** [comment]
Can you please these [images](https://dl.dropboxusercontent.com/u/15234/tray.zip) for now and we'll deal with the exact resolutions later on..

**Comment by ConorIA** [comment]
Done! 😄

**Comment by lipis** [comment]
As for the tests it's failing because the `tray.createTrayIcon()` is never called and it's a bigger thing ATM.. For now just to fix the test and have it green.. update the code in `tray.js` to something like that:

```js
  useDefaultIcon: function() {
    if (appIcon == null) return;
    appIcon.setImage(iconPath);
  },

  useBadgeIcon: function() {
    if (appIcon == null) return;
    appIcon.setImage(iconBadgePath);
  }
```

**Comment by lipis** [comment]
Also.. did you sign our [SLA](https://github.com/wireapp/wire#contributing-to-the-code)?

**Comment by lipis** [comment]
Let's see if Travis is going to wake up today! :)

Can you post a screenshot of the two icons from your machine?

**Comment by ConorIA** [comment]
Here is the icon with no messages. I'm not very popular today, but I will send an image with the badge if someone writes to me. 😉 

screenshot from 2016-09-06 16-48-18

**Comment by david-wire** [comment]
CLA approved

**Comment by lipis** [comment]
Boom..

**Pull request merged**